types/devicefarmer__stf-syrup/index.d.ts

```diff
@@ -0,0 +1,57 @@
+import Bluebird = require("bluebird");
+
+type extractDesRet<RetT> = RetT extends SyrupI<never, never, infer RetX> ? RetX
+    : unknown;
+type extractBluebirdReturnR<RetT> = RetT extends Bluebird<infer RetX> ? RetX
+    : RetT;
+declare class SyrupI<
+    OptionsT extends object = any, // TODO: find a way to remove any. Maybe we union all the options that are needed for each dependency?
+    DepsT extends SyrupI[] = [],
+    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
+    RetT = unknown | void,
+    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
+    DepsRetsT extends (unknown | void)[] = [], // TODO: maybe we can extract DepsRetsT somehow?
+> {
+    constructor(options: OptionsT | null);
+    define<
+        BodyT extends (options: OptionsT, ...deps: DepsRetsT) => unknown,
+    >(
+        body: BodyT,
+    ): SyrupI<
+        OptionsT,
+        DepsT,
+        extractBluebirdReturnR<ReturnType<typeof body>>,
+        DepsRetsT
+    >;
+    dependency<DepT extends SyrupI<never, never>>(
+        dep: DepT,
+    ): SyrupI<
+        OptionsT,
+        [...DepsT, DepT],
+        RetT,
+        [...DepsRetsT, extractDesRet<DepT>]
+    >;
+    consume<NewOptionsT extends OptionsT>(
+        overrides: NewOptionsT,
```

Comment:
It's suspicious that `NewOptionsT` only appears once here. Why not just put `OptionsT` as the argument? Does this fail for some reason?

Reply:
Hm, I thought of letting users redefine and use new options. But just `OptionsT` works fine